### ISSUE TITLE
`CoulombMatrix(permutation="sorted_l2")` is not symmetric

### ISSUE THREAD

**Issue by MomoLangenstein** [SINGROUP/dscribe#89]
When using the `dscribe` Python library, v1.2.1, Coulomb matrices generated using the `sorted_l2` permutation are no longer symmetric matrices. I compared dscribe's result with a manual implementation of the sorting based on [doi:10.1002/qua.24954](https://doi.org/10.1002/qua.24954) to check that the algorithm in general works.

This manual implementation matches https://github.com/SINGROUP/dscribe/blob/79a13939d66bdc858865dc050b91be9debd3c06a/dscribe/descriptors/matrixdescriptor.py#L143-L159

Therefore, I'd guess that something might be wrong with the C++ implementation in https://github.com/SINGROUP/dscribe/blob/79a13939d66bdc858865dc050b91be9debd3c06a/dscribe/ext/cm.cpp#L111-L136

Below is an example to reproduce the issue:

```python
import ase
import numpy as np

from io import StringIO

from ase.io import read
from dscribe.descriptors import CoulombMatrix


# Manual implementation of the CM matrix generation
def generate_coulomb_matrix(n_atoms_max, atoms):
    cm = np.zeros((n_atoms_max, n_atoms_max))

    Z = atoms.get_atomic_numbers()
    xyz = np.array(atoms.positions)

    for r in range(len(xyz)):
        for c in range(len(xyz)):
            if r == c:
                cm[r][c] = 0.5 * Z[r] ** 2.4
            elif r < c:
                cm[r][c] = Z[r] * Z[c] / np.linalg.norm(xyz[r] - xyz[c])
            else:
                cm[r][c] = cm[c][r]

    return cm

# Manual implementation of the L2 sort for a CM
def sort_cm_l2(cm):
    """
    [Coulomb matrices are] not invariant to reindexing [their] atoms.
    One remedy is to sort Coulomb matrices by descending row norm by
    simultaneously permuting rows and columns accordingly. This
    ensures invariance against nuclear permutations, but introduces
    discontinuities where the sorting changes.

    Rupp, M. (2015) Machine learning for quantum mechanics in a nutshell.
    International Journal of Quantum Chemistry.
    115 (16) pp.1058–1073. doi:10.1002/qua.24954.
    """

    row_norms = np.linalg.norm(cm, axis=1)
    sorted_row_indices = np.argsort(row_norms)[::-1]

    sorted_cm = cm[sorted_row_indices]
    sorted_cm = sorted_cm[:, sorted_row_indices]

    return sorted_cm


# Example xyz file for the pyridine molecule
# https://en.wikipedia.org/w/index.php?title=XYZ_file_format&oldid=1002382601#Example
xyz = """11

C       -0.180226841      0.360945118     -1.120304970
C       -0.180226841      1.559292118     -0.407860970
C       -0.180226841      1.503191118      0.986935030
N       -0.180226841      0.360945118      1.29018350
C       -0.180226841     -0.781300882      0.986935030
C       -0.180226841     -0.837401882     -0.407860970
H       -0.180226841      0.360945118     -2.206546970
H       -0.180226841      2.517950118     -0.917077970
H       -0.180226841      2.421289118      1.572099030
H       -0.180226841     -1.699398882      1.572099030
H       -0.180226841     -1.796059882     -0.917077970
"""


atoms = read(StringIO(xyz), format="xyz")


cm_none = CoulombMatrix(n_atoms_max=11, permutation="none", flatten=False)
cm_l2 = CoulombMatrix(n_atoms_max=11, permutation="sorted_l2", flatten=False)


dscribe_cm_none = cm_none.create(atoms).round(1)
manual_cm_none = generate_coulomb_matrix(11, atoms).round(1)

# Succeeds, i.e. the CM matrix is generated as expected
assert np.array_equal(dscribe_cm_none, manual_cm_none)


dscribe_cm_l2 = cm_l2.create(atoms).round(1)
manual_cm_l2 = sort_cm_l2(generate_coulomb_matrix(11, atoms)).round(1)

# Fails, i.e. the L2-sorted CM matrix is NOT generated as expected
assert np.array_equal(dscribe_cm_l2, manual_cm_l2)


dscribe_cm_norm = np.linalg.norm(dscribe_cm_l2, axis=0).round(1)
dscribe_cm_norm = np.linalg.norm(manual_cm_l2, axis=0).round(1)

# Succeeds, i.e. the L2-sorted CM matrix is sorted by each row's L2 norm
assert np.array_equal(dscribe_cm_none, manual_cm_none)


# Succeeds, i.e. the manually sorted CM remains symmetric
assert np.array_equal(manual_cm_none, manual_cm_none.T)


# Fails, i.e. dscribe's L2-sorted CM is no longer symmetric
assert np.array_equal(dscribe_cm_l2, dscribe_cm_l2.T)
```

**Comment by lauri-codes** [comment]
Hi @MomoLangenstein,

Thanks for the report. I need to check what is going on. If this is the case, we definitely need to fix it and introduce a better test that also checks the symmetricity.

**Comment by lauri-codes** [comment]
Hi @MomoLangenstein,

Indeed there was an issue where the rows and columns of the matrix were not being permuted as expected, causing non-symmetric matrices. There is now a fix in the branch v1.2.2, which will be made into a release in the near future.

It would be great if you can verify this fix as well. I have added a new symmetricity test into our test suite and your examples now also pass after changing the sorting on the python side slightly like this:

```python
sorted_row_indices = np.argsort(-row_norms)
```

The sorting in your manual implementation needs to be exactly the same as in our C++ implementation due to ambiguity in sorting by row norm. In several cases the norms of the matrix rows are identical and in order to get the exact same results, the sorting logic needs to be identical in your example and in our C++ implementation.

**Comment by MomoLangenstein** [comment]
Hi @lauri-codes,

Thank you for working out the fix for this issue! I can confirm that my small test now works on the `v1.2.2` branch. I just wanted to clarify your proposed change to the manual Python implementation. If I do use
```python
sorted_row_indices = np.argsort(-row_norms)
```
instead of
```python
sorted_row_indices = np.argsort(row_norms)[::-1]
```
I get exactly the same results with my manual implementation as with `dscribe`. Are both methods semantically correct, and only differ in their handling of how to order rows/columns with equivalent norms?

**Comment by lauri-codes** [comment]
Hi,

Yes, the sorting simply affects the way that rows with equal norm are handled. If there are no equal norms, then the output is completely independent of the sorting algorithm (see e.g. the `kind` attribute in the [numpy docs](https://numpy.org/doc/stable/reference/generated/numpy.argsort.html)) and whether you do `np.argsort(-row_norms)` or np.argsort(row_norms)[::-1].

The small differences only start kicking in when there are equal norms. I think `np.argsort(-row_norms)` and `np.argsort(row_norms)[::-1]` should not be equal in these cases, e.g.:

```python
import numpy as np

a = [
    [2, 1],
    [1, 2],
]

row_norms = np.linalg.norm(a, axis=1)
indices1 = np.argsort(row_norms)[::-1]
print(indices1) # [1, 0]
indices2 = np.argsort(-row_norms)
print(indices2) # [0, 1]
```

In the end it does not really matter which way you sort the rows with equal norms, just as long as you are consistent throughout your data.

**Comment by lauri-codes** [comment]
Now fixed in v1.2.2.
